File: sample-videochat-webrtc/sample-videochat-webrtc/Views/ToolBar/ButtonsFactory.m
//
//  ButtonsFactory.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "ButtonsFactory.h"
#import "Button.h"

const CGRect kDefRect = {0, 0, 44, 44};
const CGRect kDefDeclineRect = {0, 0, 96, 44};
const CGRect kDefCircleDeclineRect = {0, 0, 44, 44};

#define kDefBackgroundColor [UIColor colorWithRed:0.8118 green:0.8118 blue:0.8118 alpha:1.0]
#define kDefSelectedColor [UIColor colorWithRed:0.3843 green:0.3843 blue:0.3843 alpha:1.0]
#define kDefDeclineColor [UIColor colorWithRed:0.8118 green:0.0 blue:0.0784 alpha:1.0]
#define kDefAnswerColor [UIColor colorWithRed:0.1434 green:0.7587 blue:0.1851 alpha:1.0]

@implementation ButtonsFactory

#pragma mark - Private

+ (Button *)buttonWithFrame:(CGRect)frame
              backgroundColor:(UIColor *)backgroundColor
                selectedColor:(UIColor *)selectedColor  {
    
    Button *button = [[Button alloc] initWithFrame:frame];
    button.backgroundColor = backgroundColor;
    button.selectedColor = selectedColor;
    
    return button;
}

+ (UIImageView *)iconViewWithNormalImage:(NSString *)normalImage
                           selectedImage:(NSString *)selectedImage {
    
    UIImage *icon = [UIImage imageNamed:normalImage];
    UIImage *selectedIcon = [UIImage imageNamed:selectedImage];
    
    UIImageView *iconView = [[UIImageView alloc] initWithImage:icon
                                              highlightedImage:selectedIcon];
    
    iconView.contentMode = UIViewContentModeScaleAspectFit;

    return iconView;
}

#pragma mark - Public

+ (Button *)videoEnable {
    
    Button *button = [self buttonWithFrame:kDefRect
                             backgroundColor:kDefBackgroundColor
                               selectedColor:kDefSelectedColor];
    button.pushed = YES;
    
    button.iconView = [self iconViewWithNormalImage:@"camera_on_ic"
                                      selectedImage:@"camera_off_ic"];
    return button;
}

+ (Button *)auidoEnable {
    
    Button *button = [self buttonWithFrame:kDefRect
                             backgroundColor:kDefBackgroundColor
                               selectedColor:kDefSelectedColor];
    
    button.pushed = YES;
    
    button.iconView = [self iconViewWithNormalImage:@"mute_on_ic"
                                      selectedImage:@"mute_off_ic"];
    return button;
}

+ (Button *)dynamicEnable {
    
    Button *button = [self buttonWithFrame:kDefRect
                             backgroundColor:kDefBackgroundColor
                               selectedColor:kDefSelectedColor];
    
    button.pushed = YES;
    button.iconView = [self iconViewWithNormalImage:@"ic_volume_low"
                                      selectedImage:@"ic_volume_high"];
    return button;
}

+ (Button *)screenShare {
    
    Button *button = [self buttonWithFrame:kDefRect
                             backgroundColor:kDefBackgroundColor
                               selectedColor:kDefSelectedColor];
    
    button.iconView = [self iconViewWithNormalImage:@"screensharing_ic"
                                      selectedImage:@"screensharing_ic"];
    return button;
}

+ (Button *)answer {
    
    Button *button = [self buttonWithFrame:kDefRect
                             backgroundColor:kDefAnswerColor
                               selectedColor:kDefSelectedColor];
    
    button.iconView = [self iconViewWithNormalImage:@"answer"
                                      selectedImage:@"answer"];
    return button;
}

+ (Button *)decline {
    
    Button *button = [self buttonWithFrame:kDefDeclineRect
                             backgroundColor:kDefDeclineColor
                               selectedColor:kDefSelectedColor];
    
    button.iconView = [self iconViewWithNormalImage:@"decline-ic"
                                      selectedImage:@"decline-ic"];
    return button;
}

+ (Button *)circleDecline {
    
    Button *button = [self buttonWithFrame:kDefCircleDeclineRect
                             backgroundColor:kDefDeclineColor
                               selectedColor:kDefSelectedColor];
    
    button.iconView = [self iconViewWithNormalImage:@"decline-ic"
                                      selectedImage:@"decline-ic"];
    return button;
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Call/ToolBar/ButtonsFactory.swift
//
//  ButtonsFactory.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 12/18/18.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit

struct ButtonsFactoryConstants {
    static let rect = CGRect(x: 0.0, y: 0.0, width: 44.0, height: 44.0)
    static let declineRect = CGRect(x: 0.0, y: 0.0, width: 96.0, height: 44.0)
    static let circleDeclineRect = CGRect(x: 0.0, y: 0.0, width: 44.0, height: 44.0)
    
    static let backgroundColor = UIColor(red: 0.8118, green: 0.8118, blue: 0.8118, alpha: 1.0)
    static let selectedColor = UIColor(red: 0.3843, green: 0.3843, blue: 0.3843, alpha: 1.0)
    static let declineColor = UIColor(red: 0.8118, green: 0.0, blue: 0.0784, alpha: 1.0)
    static let answerColor = UIColor(red: 0.1434, green: 0.7587, blue: 0.1851, alpha: 1.0)
}

class ButtonsFactory {
    // MARK: - Class Methods
    class func button(withFrame frame: CGRect, backgroundColor: UIColor, selectedColor: UIColor) -> CustomButton {
        let button = CustomButton(frame: frame)
        button.backgroundColor = backgroundColor
        button.selectedColor = selectedColor
        return button
    }
    
    class func iconView(withNormalImage normalImage: String, selectedImage: String) -> UIImageView? {
        let icon = UIImage(named: normalImage)
        let selectedIcon = UIImage(named: selectedImage)
        let iconView = UIImageView(image: icon, highlightedImage: selectedIcon)
        iconView.contentMode = .scaleAspectFit
        return iconView
    }
    
    class func videoEnable() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.rect, backgroundColor:ButtonsFactoryConstants.backgroundColor, selectedColor: ButtonsFactoryConstants.selectedColor)
        button.pushed = true
        button.iconView = self.iconView(withNormalImage: "camera_on_ic", selectedImage: "camera_off_ic")
        return button
    }
    
    class func audioEnable() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.rect, backgroundColor: ButtonsFactoryConstants.backgroundColor, selectedColor: ButtonsFactoryConstants.selectedColor)
        button.pushed = true
        button.iconView = self.iconView(withNormalImage: "mute_on_ic", selectedImage: "mute_off_ic")
        return button
    }
    
    class func dynamicEnable() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.rect, backgroundColor: ButtonsFactoryConstants.backgroundColor, selectedColor: ButtonsFactoryConstants.selectedColor)
        button.pushed = true
        button.iconView = self.iconView(withNormalImage: "ic_volume_low", selectedImage: "ic_volume_high")
        return button
    }
    
    class func screenShare() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.rect, backgroundColor: ButtonsFactoryConstants.backgroundColor, selectedColor: ButtonsFactoryConstants.selectedColor)
        button.iconView = self.iconView(withNormalImage: "screensharing_ic", selectedImage: "screensharing_ic")
        return button
    }
    
    class func answer() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.rect, backgroundColor: ButtonsFactoryConstants.answerColor, selectedColor: ButtonsFactoryConstants.selectedColor)
        button.iconView = self.iconView(withNormalImage: "answer", selectedImage: "answer")
        return button
    }
    
    class func decline() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.declineRect, backgroundColor: ButtonsFactoryConstants.declineColor, selectedColor: ButtonsFactoryConstants.selectedColor)
        button.iconView = self.iconView(withNormalImage: "decline-ic", selectedImage: "decline-ic")
        return button
    }
    
    class func circleDecline() -> CustomButton {
        let button: CustomButton = self.button(withFrame: ButtonsFactoryConstants.circleDeclineRect, backgroundColor: ButtonsFactoryConstants.declineColor, selectedColor: ButtonsFactoryConstants.selectedColor)
        button.iconView = self.iconView(withNormalImage: "decline-ic", selectedImage: "decline-ic")
        return button
    }
}


